Prevent deadlock between RapidsBufferStore and RapidsBufferBase on close

abellina opened the pull request:
Signed-off-by: Alessandro Bellina <abellina@nvidia.com>

Targeting this to 22.04 but I can target to 22.02 if desired.

Closes https://github.com/NVIDIA/spark-rapids/issues/4664.

Prevents a deadlock between `RapidsBufferStore` and `RapidsBufferBase` on shutdown, removing code that was added before that was causing #4664 (this patch no longer stops the context -> no longer causes shuffle id conflicts on register)

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala:
```diff
@@ -232,6 +239,8 @@ object RapidsBufferCatalog extends Logging with Arm {
       gdsStorage = null
     }
+
+    singleton.removeAllBuffersOnShutdown()
```
Wondering if this will just move the exception somewhere else.  For example, could we now get "unknown buffer ID" if an outstanding buffer gets unregistered after catalog shutdown?

It feels like we're patching a symptom rather than fixing the root cause of why all the buffers didn't get unregistered for the first context.

abellina commented:
This PR: https://github.com/NVIDIA/spark-rapids/pull/4374, added a `.close` for the `SparkSession` in the mortgage suite, saying that there was a hang, but not explaining anything more. I removed the `.close` and that uncovered a deadlock, handled in https://github.com/NVIDIA/spark-rapids/pull/4669/commits/4e302d2fdd2c8dccd611615e5725f204d1ffc664. I reverted my other change to remove all buffers because this keeps the `SparkContext` alive, and so the conflict of shuffle ids is no longer seen.

That said, I now see device memory leaks. I'll dig at that next. I am not sure if those are related to my change.

abellina commented:
Ok the reason for the memory leaks is because of the shutdown hook logic in the JVM. The shutdown hooks, as far as I understand, run in parallel, so a registered hook for the `MemoryCleaner` runs at any random time, even before Spark is able to stop it's context, unregister shuffles, etc. One approach that would remove ambiguity would be to change the `MemoryCleaner` to not hook directly to the JDK, but instead provide the API so that a custom shutdown order could be imposed (potentially registering the `MemoryCleaner` against Spark's `ShutdownHookManager`).

I think this is a nice to have but it would make the code easier to follow. It took me a while to realize that the cuDF MemoryCleaner is racing against spark's ShutdownHookManager.

abellina commented:
build

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferStore.scala:
```diff
@@ -70,11 +70,17 @@ abstract class RapidsBufferStore(
       }
```
2022 copyrights

abellina replied:
done

jlowe reviewed (commented):
Would be good to file an issue about the underlying problem that still isn't addressed, re: closing and restarting the Spark session

abellina commented:
build

abellina commented:
build

abellina commented:
It is not clear to me why the previous one failed. I am looking to get some logs out of this to see if I can figure it out.

abellina commented:
build

abellina commented:
I am going to merge this. The previous time this failed, it seemed to die at `GpuDeviceManagerSuite`. Looking at the test, it seems to be a really simple test, that is just checking values constructed in the suite. So I am not sure why it failed, and I've had successful CI with this patch a few times.
